api/v1alpha1/openstackserver_types.go

```diff
@@ -0,0 +1,182 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-openstack/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-openstack/pkg/utils/optional"
+)
+
+const (
+	// OpenStackServerFinalizer allows ReconcileOpenStackServer to clean up resources associated with OpenStackServer before
+	// removing it from the apiserver.
+	OpenStackServerFinalizer = "openstackserver.infrastructure.cluster.x-k8s.io"
+)
+
+// OpenStackServerSpec defines the desired state of OpenStackServer.
+type OpenStackServerSpec struct {
+	// AdditionalBlockDevices is a list of specifications for additional block devices to attach to the server instance.
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	AdditionalBlockDevices []infrav1.AdditionalBlockDevice `json:"additionalBlockDevices,omitempty"`
+
+	// AvailabilityZone is the availability zone in which to create the server instance.
+	//+optional
+	AvailabilityZone optional.String `json:"availabilityZone,omitempty"`
+
+	// ConfigDrive is a flag to enable config drive for the server instance.
+	// +optional
+	ConfigDrive optional.Bool `json:"configDrive,omitempty"`
+
+	// FailureDomain is the failure domain the server will be created in.
+	// Must match a key in the FailureDomains map stored on the cluster object.
+	// +optional
+	FailureDomain optional.String `json:"failureDomain,omitempty"`
+
+	// The flavor reference for the flavor for the server instance.
+	// +required
+	Flavor string `json:"flavor"`
+
+	// FloatingIPPoolRef is a reference to a FloatingIPPool to allocate a floating IP from.
+	// +optional
+	FloatingIPPoolRef *corev1.TypedLocalObjectReference `json:"floatingIPPoolRef,omitempty"`
+
+	// IdentityRef is a reference to a secret holding OpenStack credentials.
+	// +kubebuilder:validation:Required
+	IdentityRef infrav1.OpenStackIdentityReference `json:"identityRef,omitempty"`
+
+	// The image to use for the server instance.
+	// +required
+	Image infrav1.ImageParam `json:"image"`
+
+	// Ports to be attached to the server instance.
+	// +optional
+	Ports []infrav1.PortOpts `json:"ports,omitempty"`
```

Comment:
I think Ports should be required since a server can't be created without pre-defined networking in CAPO context. Before making that change, let me know what you think.